src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,191 @@
+            TextSpan span,
+            CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var nodeToFix = root.FindNode(span, getInnermostNodeForTie: true);
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            if (semanticModel.GetOperation(nodeToFix) is IInvocationOperation typedNodeToFix)
+            {
+                var generator = SyntaxGenerator.GetGenerator(document);
+
+                var fixComponents = GetFixComponents(generator, typedNodeToFix);
+
+                var (startIndexArgument, lengthArgument) = fixComponents.OriginalInnerArguments[0].Parameter.Name == "length"
+                    ? (fixComponents.OriginalInnerArguments[1], fixComponents.OriginalInnerArguments[0])
+                    : (fixComponents.OriginalInnerArguments[0], fixComponents.OriginalInnerArguments[1]);
+                
+                // from sb.Append(text.Substring(2, 5)) generate sb.Append(text, 2, 5)
+                var newNode = generator.InvocationExpression(
+                    fixComponents.TargetMethod,
+                    fixComponents.StringArgument.Syntax,
+                    startIndexArgument.Value.Syntax,
+                    lengthArgument.Value.Syntax);
+                var newRoot = root.ReplaceNode(
+                    nodeToFix,
+                    newNode);
+
+                return document.WithSyntaxRoot(newRoot);
+            }
+
+            return document;
+        }
+
+        private static FixComponents GetFixComponents(
+            SyntaxGenerator generator, 
+            IInvocationOperation typedNodeToFix)
+        {
+            var argumentOperation = typedNodeToFix.Arguments[0].Value;
+            if (argumentOperation is IInvocationOperation stringBuilderAppendInvocationCandidate)
+            {
+                // if the stringBuilder instance is sb, generate sb.Append()
+                var append = generator.MemberAccessExpression(
+                    typedNodeToFix.Instance.Syntax,
+                    generator.IdentifierName(nameof(StringBuilder.Append)));
+
+                return new FixComponents(
+                    stringArgument: stringBuilderAppendInvocationCandidate.Instance,
+                    targetMethod: append,
+                    originalInnerArguments: stringBuilderAppendInvocationCandidate.Arguments);
+            }
+
+            return default(FixComponents);
+        }
+
+        // Needed for Telemetry (https://github.com/dotnet/roslyn-analyzers/issues/192)
+        private class MyCodeAction : DocumentChangeAction
+        {
+            public MyCodeAction(
+                string title, 
+                Func<CancellationToken, Task<Document>> createChangedDocument)
+                : base(
+                    title, 
+                    createChangedDocument, 
+                    title)
```

Comment:
1. this is all extremely excessive in terms of wrapping.  `base(title, createChangedDocument, title)`  is idiomatic, concise, and follows the patterns of the Roslyn codebase.  i get that you like this style.  But it really makes all this code stick out as being unnecessarily different from the rest of the repo.